### PR TITLE
Query builder options and create index for specialization of doctors

### PR DESCRIPTION
**Pull request by vladimir-paziuk**

### CODE REVIEW

File: apps/clinic/src/modules/doctors/dtos/doctor-query.dto.ts [hunk 1]
```diff
@@ -1,6 +1,19 @@
 import { ApiProperty } from '@nestjs/swagger';
 
+// Find way how to import it from SelectQueryBuilder orderBys.direction
+export type QueryBuilderSortDirection = 'ASC' | 'DESC';
+
 export class DoctorQueryDto {
   @ApiProperty({ required: false })
   search: string;
+
```
Comment (macc4) [inline apps/clinic/src/modules/doctors/dtos/doctor-query.dto.ts]: based on the repository getDoctors i've noticed that search = specialization, so it should be named as such, "search" is very vague

File: apps/clinic/src/modules/doctors/dtos/doctor-query.dto.ts [hunk 2]
```diff
@@ -1,6 +1,19 @@
...
+  @ApiProperty({ required: false })
+  offset: number;
+  @ApiProperty({ required: false })
+  limit: number;
+
+  @ApiProperty({ required: false })
+  sortBy: string; // groupBy
+  @ApiProperty({ required: false })
+  sortDirection: QueryBuilderSortDirection; // orderBy
 }
```
Comment (macc4) [inline apps/clinic/src/modules/doctors/dtos/doctor-query.dto.ts]: you should add type checking and enum validation, sortBy can be an enum too, just like sortDirection, but nothing is being validated right now, it seems

you can also add the enum in the ApiProperty, so the documentaion would be more comprehensive

File: apps/clinic/migrations/1658405108139-CreateDoctorsIndex.ts [hunk 3]
```diff
@@ -0,0 +1,14 @@
+import { MigrationInterface, QueryRunner } from 'typeorm';
+
+export class CreateDoctorsIndex1658405108139 implements MigrationInterface {
+  name = 'CreateDoctorsIndex1658405108139';
+  public async up(queryRunner: QueryRunner): Promise<void> {
+    await queryRunner.query(`
+      CREATE INDEX doctors_specialization ON clinic.doctors(specialization);
```
Comment (alexmaltsv) [inline apps/clinic/migrations/1658405108139-CreateDoctorsIndex.ts:7]: LOWER(specialization)

Comment (alexmaltsv) [inline apps/clinic/migrations/1658405108139-CreateDoctorsIndex.ts:7]: doctors_specialization_idx

File: apps/clinic/src/modules/doctors/doctors.repository.ts [hunk 4]
```diff
@@ -15,13 +15,21 @@ export class DoctorsRepository extends Repository<DoctorEntity> {
...
-      query.andWhere(
-        '(LOWER(doctor.name) LIKE LOWER(:search) OR LOWER(doctor.desc) LIKE LOWER(:search))',
-        { search: `%${search}%` },
-      );
+      query.andWhere('(LOWER(doctor.specialization) LIKE LOWER(:search))', {
+        search: `%${search}%`,
+      });
+    }
+    if (sortBy) {
+      query.orderBy(`doctor.${sortBy}`, sortDirection);
```
Comment (alexmaltsv) [inline apps/clinic/src/modules/doctors/doctors.repository.ts:26]: Index?

File: apps/clinic/src/modules/doctors/dtos/doctor-query.dto.ts [hunk 5]
```diff
@@ -1,6 +1,33 @@
 import { ApiProperty } from '@nestjs/swagger';
+import { IsEnum, IsString, IsNumberString } from 'class-validator';
+
+// Find way how to import it from SelectQueryBuilder orderBys.direction
+export enum queryBuilderSortDirection {
+  ASC = 'ASC',
+  DESC = 'DESC',
+}
+export enum queryBuilderSortBy {
+  specialization = 'specialization',
```
Comment (alexmaltsv) [inline apps/clinic/src/modules/doctors/dtos/doctor-query.dto.ts:10]: create indexes

File: apps/clinic/src/modules/doctors/doctors.repository.ts [hunk 6]
```diff
@@ -15,13 +15,21 @@ export class DoctorsRepository extends Repository<DoctorEntity> {
 
   async getDoctors(params: DoctorQueryDto): Promise<DoctorEntity[]> {
     const query = this.createQueryBuilder('doctor');
-    const { search } = params;
+    const { search, offset, limit, sortBy, sortDirection = 'ASC' } = params;
 
     if (search) {
```
Comment (alexmaltsv) [inline apps/clinic/src/modules/doctors/doctors.repository.ts:20]: Implement filters by date

**Pull request merged**